# <center> KLASIFIKASI dengan kNN <center>

## Import Dataset

In [29]:
from sklearn import datasets

In [30]:
iris = datasets.load_iris()

In [31]:
print(iris.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

In [32]:
x = iris.data

In [33]:
y = iris.target

In [34]:
x.shape

(150, 4)

-----------------------------

## Training & Testing

In [35]:
# holdout
from sklearn.model_selection import train_test_split

In [36]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, stratify=y)

## Normalisasi

In [37]:
from sklearn.preprocessing import StandardScaler

In [38]:
 scaler = StandardScaler().fit(x_train)

In [39]:
x_train = scaler.fit_transform(x_train)
x_test = scaler.fit_transform(x_test)

-----------------

## Klasifikasi 

In [40]:
# knn
from sklearn.neighbors import KNeighborsClassifier

In [41]:
# k = 3
classifier = KNeighborsClassifier(n_neighbors=3)

In [42]:
# classifier untuk belajar dari data kita
# fase training
classifier.fit(x_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [43]:
# hasil prediksi
classifier.predict(x_test)

array([2, 1, 1, 0, 0, 0, 1, 2, 2, 2, 0, 2, 0, 1, 1, 2, 0, 1, 2, 1, 1, 2,
       0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 2, 0, 2, 0, 2])

In [44]:
y_test

array([2, 1, 1, 0, 0, 0, 1, 2, 2, 2, 0, 2, 0, 1, 2, 2, 0, 1, 2, 1, 1, 2,
       0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 2, 0, 2, 0, 2])

In [45]:
# akurasi
classifier.score(x_test, y_test)

0.9736842105263158

akurasi memiliki rentang nilai 0-1

0 artinya prediksi salah semua

1 artinya prediksi benar semua

-----------------

## Eksperimen

k = [3, 5, 7, 9, 11, 13, 15, 17, 19, 21]

In [46]:
# list nilai k
k = [3, 5, 7, 9, 11, 13, 15, 17, 19, 21]

In [47]:
# untuk setiap nilai pada list k
for i in k:
    print(i)

3
5
7
9
11
13
15
17
19
21


In [48]:
for i in k:
    # set nilai k pada knn
    classifier = KNeighborsClassifier(n_neighbors=i)
    # training
    classifier.fit(x_train, y_train)
    # evaluasi/testing
    print('k = {} dengan accuracy = {}'.format(i, classifier.score(x_test, y_test)*100))

k = 3 dengan accuracy = 97.36842105263158
k = 5 dengan accuracy = 94.73684210526315
k = 7 dengan accuracy = 94.73684210526315
k = 9 dengan accuracy = 94.73684210526315
k = 11 dengan accuracy = 94.73684210526315
k = 13 dengan accuracy = 94.73684210526315
k = 15 dengan accuracy = 92.10526315789474
k = 17 dengan accuracy = 94.73684210526315
k = 19 dengan accuracy = 92.10526315789474
k = 21 dengan accuracy = 94.73684210526315


**kesimpulannya, nilai k terbaik pada contoh kasus dataset Iris adalah k = 7 dengan accuracy 97.37**